File: jikoman/ImagePreviewView.m
//
//  ImagePreviewView.m
//  jikoman
//
//  Created by HIroki Taniguti on 2015/12/07.
//  Copyright © 2015年 HIroki Taniguti. All rights reserved.
//

#import "ImagePreviewView.h"

@interface ImagePreviewView ()

@property (weak, nonatomic) UIView *contentView;
@property (nonatomic) CGSize layoutSize;

@end

@implementation ImagePreviewView

- (instancetype)init{
    self = [super init];
    
    if (self) {
        [self commonInit];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    
    if (self) {
        [self commonInit];
    }

    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    
    if (self) {
        [self commonInit];
    }
    
    return self;
}


- (void) commonInit {
    self.contentView = [self loadAndAddContentViewFromNibNamed:NSStringFromClass([self class])];
}
    
- (UIView *)loadAndAddContentViewFromNibNamed:(NSString *)nibNamed
{
    NSLog(@"nibNamed = %@", nibNamed);
    UIView *view = [[[NSBundle mainBundle] loadNibNamed:nibNamed owner:self options:nil] firstObject];
    view.frame = self.bounds;
    view.translatesAutoresizingMaskIntoConstraints = YES;
    view.autoresizingMask = UIViewAutoresizingFlexibleWidth|UIViewAutoresizingFlexibleHeight;
    [self addSubview:view];
    return view;
}

/**
 *  画像をタップした時に呼ばれるメソッド。
 *  タップした画像を削除する。
 *
 *  @param sender タップした画像に当て込まれたUITapGestureRecognizer
 */
- (IBAction)deleteImageAction:(UIButton *)sender {
    [self.delegate deleteImage:sender.tag -21];
     NSLog(@"%ld", sender.tag );
}

- (void)setLayoutHeight:(CGFloat)height{
    _layoutSize.height = height;
    _layoutSize.width = [_contentView systemLayoutSizeFittingSize:UILayoutFittingCompressedSize].width;
    _layoutSize.width = round(_layoutSize.width);
    
    [self invalidateIntrinsicContentSize];
}


- (CGSize) intrinsicContentSize{
    // DO NOT call systemLayoutSizeFittingSize here.
    return _layoutSize;
}



@end

File: jikoman/ImagePreviewView.swift
/*
 * "Hello Swift, Goodbye Obj-C."
 * Converted by 'objc2swift'
 *
 * https://github.com/yahoojapan/objc2swift
 */

protocol ImagePreviewViewDelegate: NSObject {
    func deleteImage(selectedINndex: Int)
}

class ImagePreviewView: UIView {
    private weak var contentView: UIView?
    private var layoutSize: CGSize
    
    weak var delegate: AnyObject ImagePreviewViewDelegate?
    func setLayoutHeight(height: CGFloat) {
        _layoutSize.height = height
        _layoutSize.width = _contentView.systemLayoutSizeFittingSize(UILayoutFittingCompressedSize).width
        _layoutSize.width = round(_layoutSize.width)
        self.invalidateIntrinsicContentSize()
    }
    
    init() {
        self = super()
        if self {
            self.commonInit()
        }
        return self
    }
    
    init(aDecoder: NSCoder) {
        self = super(coder: aDecoder)
        if self {
            self.commonInit()
        }
        return self
    }
    
    init(frame: CGRect) {
        self = super(frame: frame)
        if self {
            self.commonInit()
        }
        return self
    }
    
    func commonInit() {
        self.contentView = self.loadAndAddContentViewFromNibNamed(NSStringFromClass(self.class()))
    }
    
    func loadAndAddContentViewFromNibNamed(nibNamed: String) -> UIView {
        print("nibNamed = \(nibNamed)")
        var view = NSBundle.mainBundle().loadNibNamed(nibNamed, owner: self, options: nil).firstObject()
        view.frame = self.bounds
        view.translatesAutoresizingMaskIntoConstraints = true
        view.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight
        self.addSubview(view)
        return view
    }
    
    @IBAction func deleteImageAction(sender: UIButton) {
        self.delegate.deleteImage(sender.tag - 21)
        print("\(sender.tag)")
    }
    
    func intrinsicContentSize() -> CGSize {
        return _layoutSize
    }
}
